crates/common/consensus/lean/src/state.rs

```diff
@@ -48,126 +59,362 @@ impl LeanState {
+    fn process_block(&mut self, block: &Block) -> anyhow::Result<()> {
+        // Send latest head slot to metrics
+        set_int_gauge_vec(&HEAD_SLOT, block.slot as i64, &[]);
+
+        self.process_block_header(block)?;
+        self.process_operations(&block.body)?;
+
+        Ok(())
+    }
+
+    fn process_block_header(&mut self, block: &Block) -> anyhow::Result<()> {
+        // Verify that the slots match
+        ensure!(
+            block.slot == self.slot,
+            "Block slot number does not match state slot number"
+        );
+        // Verify that the block is newer than latest block header
+        ensure!(
+            block.slot > self.latest_block_header.slot,
+            "Block slot number is not greater than latest block header slot number"
+        );
+        // Verify that the proposer index is the correct index
+        ensure!(
+            block.proposer_index == block.slot % self.config.num_validators,
+            "Block proposer index does not match the expected proposer index"
+        );
+
+        // Verify that the parent matches
+        ensure!(
+            block.parent_root == self.latest_block_header.tree_hash_root(),
+            "Block parent root does not match latest block header root"
+        );
+
+        // If this was first block post genesis, 3sf mini special treatment is required
+        // to correctly set genesis block root as already justified and finalized.
+        // This is not possible at the time of genesis state generation and are set at
+        // zero bytes because genesis block is calculated using genesis state causing a
+        // circular dependancy
+        if self.latest_block_header.slot == 0 {
+            // block.parent_root is the genesis root
+            self.latest_justified.root = block.parent_root;
+            self.latest_finalized.root = block.parent_root;
+        }
+
+        // now that we can vote on parent, push it at its correct slot index in the structures
+        self.historical_block_hashes
+            .push(block.parent_root)
+            .map_err(|err| {
+                anyhow!("Failed to add block.parent_root to historical_block_hashes: {err:?}")
             })?;
 
-        // Take left side of the list (if any)
-        for (i, justification_bit) in self
-            .justifications_roots_validators
-            .iter()
-            .take(index * VALIDATOR_REGISTRY_LIMIT as usize)
-            .enumerate()
-        {
-            new_justifications_roots_validators
-                .set(i, justification_bit)
-                .map_err(|err| anyhow!("Failed to set new justification bit: {err:?}"))?;
+        // genesis block is always justified
+        let is_justified = self.latest_block_header.slot == 0;
+
+        self.justified_slots
+            .push(is_justified)
+            .map_err(|err| anyhow!("Failed to add to justified_slots: {err:?}"))?;
+
```

Comment:
```suggestion
        // genesis block is always justified
        self.justified_slots
            .push(self.latest_block_header.slot == 0;)
            .map_err(|err| anyhow!("Failed to add to justified_slots: {err:?}"))?;

```

Hmm I'm afraid `is_justified` is not that descriptive here, so would suggest to inline it,

Reply:
Fixed in fc4a09f